docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,295 @@
+
+### Requirements
+
+#### Functional Requirements
+
+##### FR1. CAPV MUST support credentials provided through a `VsphereAccount` and referenced by `VsphereCluster.AccountRef` field
+
+##### FR2. CAPV MUST support static credentials
+
+##### FR3. CAPV MUST support clusterctl move scenarios
+
+##### FR4. CAPV MUST prevent privilege escalation allowing users to create clusters in accounts they should not be able to
+
+#### Non-Functional Requirements
+
+##### NFR1. Unit tests MUST exist for cluster and machine controllers that utilize the credentials
+
+##### NFR2. e2e tests MUST exist for multi account scenarios
+
+### Implementation Details/Notes/Constraints
+
+#### Current State
+
+CAPV currently uses a session manager to create and cache sessions. Sessions are created (or retrieved from the cache) during cluster and vsphere VM reconcile loops.
+vSphere cluster uses the session as a sanity check to ensure connectivity. vSphere VM uses the session for VM lifecycle tasks on vcenter and the session is stored in a VMContext.
+
+```go
+type VMContext struct {
+    *ControllerContext
+    VSphereVM   *v1alpha3.VSphereVM
+    PatchHelper *patch.Helper
+    Logger      logr.Logger
+    Session     *session.Session
+}
+```
+
+#### Proposed Changes
+
+The proposed changes below allow users to specify the account to use for clusters during runtime.
+
+Changed Resources
+
+- `VSphereCluster`
+
+New Resources
+
+- cluster-scoped `VSphereAccount` represents the vCenter account to use for reconciliation. This type should also contain a string list of namespaces in which it is allowed to be used.
+
+Changes to VSphereCluster
+A new field is added to the VSphereClusterSpec to reference the `VSphereAccount`. We intend to use `corev1.LocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VSphereAccount is the account to be used for vcenter actions
+type VSphereAccount struct {
+    metav1.TypeMeta `json:`",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec VSphereAccountSpec `json:"spec,omitempty"`
+    Status VSphereAccountStatus `json:"status, omitempty"`
+}
+
+type VSphereAccountSpec struct {
+    // Server is the address of the vSphere endpoint.
+    Server string `json:"server,omitempty"`
+
+    Username string `json:"username,omitempty"`
+
+    Password string `json:"password,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using an array of namespaces or with a label selector.
+    // If this object is nil, no namespaces will be allowed
```

Comment:
are other providers failing closed too ?

Reply:
AWS is but Azure uses a string slice and if it's empty, will default to all namespaces are allowed.

Reply:
I'm keep track of this to discuss with capz community

Reply:
@randomvariable @sedefsavas -  IIRC CAPA also has namespaced resources for account management right ? so that might be a stop-gap until clusterctl supports the cluster-scoped ones

Reply:
Yes, the idea that you can enable both the cloud admin persona to set up cluster scoped resources for use by different teams, and then the devops and developer personas can also plug in their own dev environments using the namespace scoped resources.

Reply:
@gab-satchi - this should be a stop-gap until we know how to move cluster-scoped resources